modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -113,86 +126,128 @@ CrystalPlasticityStressUpdateBase::getSlipSystems()
+    paramError(
+        "number_slip_systems",
+        "The number of rows in the slip system file should match the number of slip system.");
+
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    // initialize to zero
+    _slip_direction[i].resize(LIBMESH_DIM);
+    _slip_plane_normal[i].resize(LIBMESH_DIM);
+  }
+
+  if (_unit_cell_type == UnitCellType::HCP)
+    transformHexagonalMillerBravisSlipSystems(_reader);
+  else if (_unit_cell_type == UnitCellType::BCC || _unit_cell_type == UnitCellType::FCC)
+  {
+    for (unsigned int i = 0; i < _number_slip_systems; ++i)
+    {
+      // directly grab the raw data and scale it by the unit cell dimension
+      for (unsigned int j = 0; j < _reader.getData(i).size(); ++j)
+      {
+        if (j < LIBMESH_DIM)
+          _slip_plane_normal[i](j) = _reader.getData(i)[j] / _unit_cell_dimension[j];
+        else
+          _slip_direction[i](j - LIBMESH_DIM) =
+              _reader.getData(i)[j] * _unit_cell_dimension[j - LIBMESH_DIM];
+      }
+    }
+  }
+
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    // normalize
+    _slip_plane_normal[i].scale(1.0 / _slip_plane_normal[i].l2_norm());
+    _slip_direction[i].scale(1.0 / _slip_direction[i].l2_norm());
+
+    if (_unit_cell_type != UnitCellType::HCP)
+    {
+      // check if slip direction is normal to the slip plane normal
+      auto magnitude = _slip_plane_normal[i].dot(_slip_direction[i]);
+      if (std::abs(magnitude) > libMesh::TOLERANCE)
+      {
+        orthonormal_error = true;
+        break;
+      }
+    }
+  }
 
   if (orthonormal_error)
     mooseError("CrystalPlasticityStressUpdateBase Error: The slip system file contains a slip "
-               "direction and "
-               "plane normal pair that are not orthonormal");
+               "direction and plane normal pair that are not orthonormal in the Cartesian "
+               "coordinate system.");
 }
 
 void
-CrystalPlasticityStressUpdateBase::getPlaneNormalAndDirectionVectors(
-    const FileName & vector_file_name,
-    const unsigned int & number_dislocation_systems,
-    DenseVector<Real> & plane_normal_vector,
-    DenseVector<Real> & direction_vector,
-    bool & orthonormal_error)
+CrystalPlasticityStressUpdateBase::transformHexagonalMillerBravisSlipSystems(const MooseUtils::DelimitedFileReader & reader)
 {
-  mooseAssert(
-      LIBMESH_DIM == 3,
-      "Crystal plasticity slip plane normal and slip direction needs to have sizes equal to 3");
-
-  Real vec[LIBMESH_DIM];
-  std::ifstream fileslipsys;
+  const unsigned int miller_bravis_indices = 4;
+  std::vector<Real> miller_bravis_slip_direction, temporary_slip_direction, temporary_slip_plane;
+  miller_bravis_slip_direction.resize(miller_bravis_indices);
+  temporary_slip_plane.resize(LIBMESH_DIM);
+  temporary_slip_direction.resize(LIBMESH_DIM);
+
+  if (_unit_cell_dimension[0] != _unit_cell_dimension[1] ||
+      _unit_cell_dimension[0] == _unit_cell_dimension[2])
```

Comment:
Is it not possible at all to have `unit_cell_height=hexagon_side_length`?

Reply:
I don't think so, but I will do the close packing geometry math and post here

Reply:
yes, I think it is reasonable to say that the height (c for HCP) should not be equivalent to the side (a). The ideal c/a ratio is 1.633, and the element with the lowest c/a ratio is Be with 1.568